app/src/main/java/io/mochahub/powermeter/workoutsession/workoutsessiondialog/NewWorkoutViewModel.kt

```diff
@@ -0,0 +1,123 @@
+package io.mochahub.powermeter.workoutsession.workoutsessiondialog
+
+import android.util.Log
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import androidx.room.withTransaction
+import io.mochahub.powermeter.data.AppDatabase
+import io.mochahub.powermeter.data.ExerciseEntity
+import io.mochahub.powermeter.data.WorkoutEntity
+import io.mochahub.powermeter.data.WorkoutSessionEntity
+import io.mochahub.powermeter.data.WorkoutSetEntity
+import io.mochahub.powermeter.models.Exercise
+import io.mochahub.powermeter.models.Workout
+import io.mochahub.powermeter.models.WorkoutSession
+import io.mochahub.powermeter.models.WorkoutSet
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.launch
+import java.text.SimpleDateFormat
+import java.time.Instant
+import java.util.Locale
+
+class NewWorkoutViewModel(val db: AppDatabase) : ViewModel() {
+    var workouts = ArrayList<Workout>()
+
+    val simpleDateFormat = SimpleDateFormat("MM/dd/yy", Locale.US)
+
+    fun getExercises(): List<ExerciseEntity> {
+        return db.exerciseDao().getAllAsList()
+    }
+
+    fun getWorkouts(workoutSessionID: String): ArrayList<Workout> {
+        val workouts = ArrayList<Workout>()
+        val workoutEntities = db.workoutDao().getWorkoutsByWorkoutSession(workoutSessionID)
+
+        workoutEntities.forEach { workoutEntity ->
+            val exercise = db.exerciseDao().findByID(workoutEntity.exerciseUUID)
+            val workout = Workout(
+                exercise = Exercise(
+                    name = exercise.name,
+                    personalRecord = exercise.personalRecord,
+                    muscleGroup = exercise.muscleGroup),
+                sets = ArrayList())
+
+            val workoutSets = db.workoutSetDao().getWorkoutSetsByWorkout(workoutEntity.id)
+            workoutSets.forEach { workoutSetEntity ->
+                val workoutSet = WorkoutSet(
+                    weight = workoutSetEntity.weight,
+                    reps = workoutSetEntity.reps)
+                workout.sets.add(workoutSet)
+            }
+            workouts.add(workout)
+        }
+        return workouts
+    }
+
+    fun deleteWorkoutSession(workoutSessionID: String) {
+        db.workoutSessionDao().deleteByID(workoutSessionID = workoutSessionID)
+    }
+
+    suspend fun saveWorkoutSession(workoutSession: WorkoutSession): String? {
```

Comment:
This doesn't need `suspend` since it has its own Coroutine Builder with the `viewModelScope.launch`

Reply:
We can't use `viewModelScope.launch` because that is tied to the main thread, and database calls are not permitted in the main thread.

Reply:
Removed the `suspend` and put the logic within a `CoroutineScope(Dispatchers.IO).launch{}`. A little messy, but refactors coming soon. Changes here https://github.com/power-meter/android-client/pull/32/commits/edcd086f92847af94fe63e56d45f832c140f67e4